### ISSUE TITLE
How to compile with -Dskip_frames=false?

### ISSUE THREAD

**Issue by BaconCatBug** [Genymobile/scrcpy#563]
I saw on https://github.com/Genymobile/scrcpy/issues/244 there is a flag to disable the skipping of frames. However, when I use the command ` meson x -Dskip_frames=false ` and then compile, the resulting exe still skips frames. I also tried `cd x` and then `ninja` as well as `meson x --reconfigure` to no avail. Am I missing something obvious or does that no longer work? I am trying to compile the latest dev build for the lost focus bug fix if that makes a difference.

**Comment by rom1v** [comment]
> the resulting exe still skips frames

How do you see that? If you show fps with Ctrl+i, does it still print "skipped frames" (sometimes)?

**Comment by BaconCatBug** [comment]
Yes, it still shows up as (+2 frames skipped).

**Comment by BaconCatBug** [comment]
Do I need to do anything specific after running `meson x -Dskip_frames=false`

**Comment by rom1v** [comment]
@BaconCatBug 
Just `ninja -Cx` then `./run x` to execute.

**Comment by BaconCatBug** [comment]
Ok I just compiled the latest dev build and I am still getting the +x Frames Skipped message when I press Control+i

I am running on windows and cross compiling on my Ubuntu laptop, both the 32 and 64 bit ones have the skipped frames.

**Comment by rom1v** [comment]
This is very weird, I cannot reproduce.

If I pass `-Dskip_frames=true` or nothing, I have `(+x frames skipped)`. Otherwise, I don't.

What is the result of `meson configure x`?

**Comment by BaconCatBug** [comment]
https://pastebin.com/1gmBDF1n

**Comment by BaconCatBug** [comment]
https://pastebin.com/0ZbxBrHn

**Comment by rom1v** [comment]
```
cat x/app/config.h
```
?

**Comment by BaconCatBug** [comment]
https://pastebin.com/0PxUUVdV

**Comment by rom1v** [comment]
```c
#undef SKIP_FRAMES
```

`config.h` is correctly included in `fps_counter.h`:
https://github.com/Genymobile/scrcpy/blob/8c8649cfcd710859ce18eab557ed2af8cedb9a42/app/src/fps_counter.h#L7

The code which prints `(+x frames skipped)` is in a `#ifdef SKIP_FRAMES` block:
https://github.com/Genymobile/scrcpy/blob/8c8649cfcd710859ce18eab557ed2af8cedb9a42/app/src/fps_counter.c#L31-L36

So it should not be executed (or even compiled) in your binary. The only reason I can think of is that you don't execute the correct binary. What command do you type to execute?

**Comment by BaconCatBug** [comment]
I run `scrcpy -c 1080:1920:0:120 -m 1024 -b 20M` on Windows 8.1

**Comment by rom1v** [comment]
> after compiling with `make -f Makefile.CrossWindows`

This generates the release in `dist/`.

**Comment by BaconCatBug** [comment]
Yes, I know. I then transfer the zip to my windows machine and extract it, then run scrcpy.exe

**Comment by rom1v** [comment]
From the right directory? Do you have another `scrcpy.exe` somewhere accessible from your `PATH`?

What if you execute `.\scrcpy.exe` from your release directory?

**Comment by BaconCatBug** [comment]
I do not have any other scrcpy.exe in my path (scrcpy is not recognised).
Using the exe from the scrcpy-win64 folder instead of the zip is the same result, skipped frames.
The linux version works, no skipped frames.

**Comment by rom1v** [comment]
Oh, ok, I get it. `meson x`/`ninja -Cx` is just for your local build.

`Makefile.CrossWindows` generates the release with its own build dir and configuration.

```diff
diff --git a/Makefile.CrossWindows b/Makefile.CrossWindows
index 0f7b14a..1e59969 100644
--- a/Makefile.CrossWindows
+++ b/Makefile.CrossWindows
@@ -57,6 +57,7 @@ build-win32: prepare-deps-win32
                        --buildtype release --strip -Db_lto=true \
                        -Dcrossbuild_windows=true \
                        -Dbuild_server=false \
+                       -Dskip_frames=false \
                        -Doverride_server_path=scrcpy-server.jar )
        ninja -C "$(WIN32_BUILD_DIR)"
 
@@ -67,6 +68,7 @@ build-win32-noconsole: prepare-deps-win32
                        --buildtype release --strip -Db_lto=true \
                        -Dcrossbuild_windows=true \
                        -Dbuild_server=false \
+                       -Dskip_frames=false \
                        -Dwindows_noconsole=true \
                        -Doverride_server_path=scrcpy-server.jar )
        ninja -C "$(WIN32_NOCONSOLE_BUILD_DIR)"
@@ -81,6 +83,7 @@ build-win64: prepare-deps-win64
                        --buildtype release --strip -Db_lto=true \
                        -Dcrossbuild_windows=true \
                        -Dbuild_server=false \
+                       -Dskip_frames=false \
                        -Doverride_server_path=scrcpy-server.jar )
        ninja -C "$(WIN64_BUILD_DIR)"
 
@@ -91,6 +94,7 @@ build-win64-noconsole: prepare-deps-win64
                        --buildtype release --strip -Db_lto=true \
                        -Dcrossbuild_windows=true \
                        -Dbuild_server=false \
+                       -Dskip_frames=false \
                        -Dwindows_noconsole=true \
                        -Doverride_server_path=scrcpy-server.jar )
        ninja -C "$(WIN64_NOCONSOLE_BUILD_DIR)"
```

**Comment by BaconCatBug** [comment]
Ah, thank you, 

Adding those four `-Dskip_frames=false \` to the Makefile.CrossWindows file has fixed the issue, it is now no longer skipping frames on Windows.

Thank you for your help and apologies for taking up your time!

**Comment by rom1v** [comment]
You're welcome.

Do you see any difference in practice (except in the console)?

**Comment by BaconCatBug** [comment]
I do but it could be placebo for all I know. I'd need to do a proper side
by side with recording to be sure. Even so it's still a little choppier
than when I view the image on the phone but that might be down to my pretty
old CPU/GPU combo.

**Comment by rom1v** [comment]
How many frames were skipped each second with skip_frames enabled?

**Comment by BaconCatBug** [comment]
Between 1 and 5, with the occasional 6 or 7.

**Comment by rom1v** [comment]
On `dev` branch, I replaced the compilation flag `skip_frames` by a runtime command-line parameter:

```bash
# equivalent to the old compilation option skip_frames=false
scrcpy --render-expired-frames
```

**Comment by Xarizzar** [comment]
Apologies for posting on a closed issue, but by replaced, you mean that if we're to run scrcpy without frame skips, we cannot compile it with the previous `-Dskip_frames=false` flag, so that it runs by default without frame skipping, and we absolutely have to use the new runtime command-line parameter?

I noticed that `-Dskip_frames=false` is a flag that's passed to mason, so I just edited the makefile to include `-Dskip-frames=false` on the build-win32/64 rules, including the no console ones, for a total of 4 times, to no avail, and then I found this thread. It would appear from the posts above that you had done something similar in an older version of scrcpy. I have pasted the makefile in a pastebin, if it helps: https://pastebin.com/KrHZpfSt

If we can only stop the frame skipping using that parameter, is it possible to make it so that it can read some sort of configuration file so that I don't have to run it from the command line every time?

Thank you for your time.
